Transition 1 (nice diffs) Go, C++ and General generators with class

Lakedaemon opened the pull request:
Hopefully I got it right this time...

gwvo commented on include/flatbuffers/code_generators.h:
```diff
@@ -0,0 +1,64 @@
+   holding the parser, a path and a filename being processed.
+   Still, it brings a common structure (normalization) among generators
+
+   The many advantages of object based generators will come later from   :
```
don't make promises for later, description of these features can be in later commits.

Lakedaemon replied:
ok, it makes sense

gwvo commented on include/flatbuffers/code_generators.h:
```diff
@@ -0,0 +1,64 @@
+namespace flatbuffers {
+
+class BaseGenerator {
+public:
```
several things here not adhering to google style, e.g. indentation of public, _ in constructor arguments (they need to be on the instance variables)

Lakedaemon replied:
Do you really insist on the _ on instance variables ? 

This means that every code generator writer will have to use *countless times* parser_, file_name_, namespace_dir_, path_  in their code 
instead of parser, file_name, namespace_dir, path

But the constructor, that is used just once will be pretty.

gwvo commented on src/idl_gen_cpp.cpp:
```diff
@@ -708,10 +709,20 @@ struct IsAlnum {
-// and tables) and output them to a single file.
-std::string GenerateCPP(const Parser &parser,
-                        const std::string &file_name) {
+static std::string GeneratedFileName(const std::string &path,
```
Why was this function added? I see it only called once.

Lakedaemon replied:
It hasn't be added. It currently lives at lines 897-900 in your current code base in the idl_gen_cpp.cpp file

It has just been moved before the class because it is used inside it.

gwvo commented on src/idl_gen_cpp.cpp:
```diff
@@ -708,10 +709,20 @@ struct IsAlnum {
+  // and tables) and output them to a single file.
+  bool generate() {
   // Check if we have any code to generate at all, to avoid an empty header.
   for (auto it = parser.enums_.vec.begin(); it != parser.enums_.vec.end();
```
indent?

Lakedaemon replied:
I cut this code and pasted it there (this is moved old code). 
As I didn't want clang-format to mess this old code, so that git could correctly display was has changed and what hasn't (it still doesn't look like I thought it would look though).

I'll indent it with clang-format then (everything should be indented 2 spaces to the right as this function was moved inside a class) but I fear that it will mess things for the pr. :/

gwvo commented on src/idl_gen_general.cpp:
```diff
@@ -1146,11 +1147,18 @@ static bool SaveClass(const LanguageParameters &lang, const Parser &parser,
-                     const std::string & file_name) {
-
-  assert(parser.opts.lang <= IDLOptions::kMAX);
+/** it'll be split later in java/csharp... and moved to separate files */
```
don't put these kinds of explanations in code.. put them in the commit message or discuss them elsewhere

Lakedaemon replied:
Yes. You are right

Lakedaemon commented:
I managed to implement a short macro for jedit that calls clang-format on selected text for me. 
So, I'll clang-format all code that I paste/move/write and there should not be problems with indent anymore (hopefully)

Lakedaemon commented:
Hello. I submitted the aforementioned changes.
Could you tell me if there are still things wrong with the latest commit or if you want more changes ?

gwvo commented on src/idl_gen_cpp.cpp:
```diff
@@ -708,197 +709,205 @@ struct IsAlnum {
-    auto code = GenerateCPP(parser, file_name);
-    return !code.length() ||
-           SaveFile(GeneratedFileName(path, file_name).c_str(), code, false);
+  cpp::CppGenerator *generator = new cpp::CppGenerator(parser, path, file_name);
```
Memory leak? Where is this deleted?
Maybe just allocate it on the stack?

Lakedaemon replied:
Yes, sorry about that. I'll fix it as soon as possible.

gwvo commented:
Ok, this is indeed a confusing diff, easier to learn to apply the correct formatting yourself :) But ok, the change seems simple, as the only thing that is really different is putting the function in a class.

Looks good, but why not also do this change for the remaining languages? Python / JS / PHP.

Lakedaemon commented:
I already did the changes for the other languages in my local repository. I was waiting for your approval to push them upstream. Will it be a problem for merging if I push them now ?

The following changes should be simple too :
mostly refactoring very long methods into manageable simple short methods (as per the Google C++ style) and starting to share code among the various code generators.

Lakedaemon commented:
There. I fixed the memory leaks (didn't know how to instantiate a class on the stack before, this is quite different from java).

Also, I added js/python/php code generators to the refactoring process.
doing sh generate_code.sh is enough to check that all is well ? as well as flattests ?

gwvo commented:
yes, `generate_code.sh` would cause any of the generated code files to show up as changed in your commit if you had broken anything about the code generation process.

Thanks!
